File: scaffold/src/main/java/top/xuqingquan/web/JsInterfaceHolderImpl.java
package top.xuqingquan.web;

import android.annotation.SuppressLint;
import android.webkit.WebView;
import top.xuqingquan.utils.Timber;

import java.util.Map;
import java.util.Set;

public class JsInterfaceHolderImpl extends JsBaseInterfaceHolder {

    private WebView mWebView;

    static JsInterfaceHolderImpl getJsInterfaceHolder(WebView webView) {
        return new JsInterfaceHolderImpl(webView);
    }

    JsInterfaceHolderImpl(WebView webView) {
        this.mWebView = webView;
    }

    @Override
    public JsInterfaceHolder addJavaObjects(Map<String, Object> maps) {
        Set<Map.Entry<String, Object>> sets = maps.entrySet();
        for (Map.Entry<String, Object> mEntry : sets) {
            Object v = mEntry.getValue();
            boolean t = checkObject(v);
            if (!t) {
                throw new JsInterfaceObjectException("This object has not offer method javascript to call ,please check addJavascriptInterface annotation was be added");
            } else {
                addJavaObjectDirect(mEntry.getKey(), v);
            }
        }
        return this;
    }

    @Override
    public JsInterfaceHolder addJavaObject(String k, Object v) {
        boolean t = checkObject(v);
        if (!t) {
            throw new JsInterfaceObjectException("this object has not offer method javascript to call , please check addJavascriptInterface annotation was be added");
        } else {
            addJavaObjectDirect(k, v);
        }
        return this;
    }

    @SuppressLint("JavascriptInterface")
    private JsInterfaceHolder addJavaObjectDirect(String k, Object v) {
        Timber.i("k:" + k + "  v:" + v);
        this.mWebView.addJavascriptInterface(v, k);
        return this;
    }

}

File: scaffold/src/main/java/top/xuqingquan/web/JsInterfaceHolderImpl.kt
package top.xuqingquan.web

import com.tencent.smtt.sdk.WebView
import top.xuqingquan.utils.Timber
import top.xuqingquan.web.agent.JsBaseInterfaceHolder
import top.xuqingquan.web.agent.JsInterfaceHolder
import top.xuqingquan.web.agent.JsInterfaceObjectException

class JsInterfaceHolderImpl private constructor(private val mWebView: WebView) : JsBaseInterfaceHolder() {

    override fun addJavaObjects(maps: Map<String, Any>): JsInterfaceHolder {
        val sets = maps.entries
        for ((key, v) in sets) {
            val t = checkObject(v)
            if (!t) {
                throw JsInterfaceObjectException("This object has not offer method javascript to call ,please check addJavascriptInterface annotation was be added")
            } else {
                addJavaObjectDirect(key, v)
            }
        }
        return this
    }

    override fun addJavaObject(k: String, v: Any): JsInterfaceHolder {
        return this
    }

    private fun addJavaObjectDirect(k: String, v: Any): JsInterfaceHolder {
        Timber.i("k:$k  v:$v")
        this.mWebView.addJavascriptInterface(v, k)
        return this
    }

    companion object {

        @JvmStatic
        fun getJsInterfaceHolder(webView: WebView): JsInterfaceHolderImpl {
            return JsInterfaceHolderImpl(webView)
        }
    }

}
